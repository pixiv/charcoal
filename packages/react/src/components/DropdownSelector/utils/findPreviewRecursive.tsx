import React, { ReactNode } from 'react'

/**
 * DropdownSelectorの選択中の要素をレンダリングするため、
 * 選択中のMenuItemを再帰的に探索しReactNodeを返す。
 *
 * @param children
 * @param value
 * @param values
 * @returns
 */
export function findPreviewRecursive(
  children: ReactNode,
  value: string
): ReactNode | undefined {
  const childArray = React.Children.toArray(children)
  for (let i = 0; i < childArray.length; i++) {
    const child = childArray[i]
    if (React.isValidElement(child)) {
      if ('value' in child.props) {
        const childValue = (child.props as { value: string }).value
        if (childValue === value && 'children' in child.props) {
          const children = (child.props as { children: ReactNode }).children
          return children
        }
      }
      if ('children' in child.props) {
        const children = findPreviewRecursive(
          (child.props as { children: ReactNode }).children,
          value
        )
        if (children !== undefined) {
          return children
        }
      }
    }
  }
}
