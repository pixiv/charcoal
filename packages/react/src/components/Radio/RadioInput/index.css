.charcoal-radio-input {
  appearance: none;
  display: block;
  box-sizing: border-box;

  margin: 0;
  padding: 6px;

  width: 20px;
  height: 20px;
  cursor: pointer;
  border-radius: 999999px;
  background-color: var(--charcoal-color-container-default);
  transition: 0.2s background-color, 0.2s box-shadow;
}

.charcoal-radio-input:checked {
  background-color: var(--charcoal-color-container-primary-default);
}

.charcoal-radio-input:checked::after {
  content: '';
  display: block;
  width: 8px;
  height: 8px;
  pointer-events: none;
  background-color: var(--charcoal-color-icon-on-primary-default);
  border-radius: 999999px;
  transition: 0.2s background-color, 0.2s box-shadow;
}

.charcoal-radio-input:not(:checked) {
  border-width: 2px;
  border-style: solid;
  border-color: var(--charcoal-color-text-tertiary-default);
}

.charcoal-radio-input:disabled {
  cursor: default;
}

.charcoal-radio-input:not(:disabled):hover {
  background-color: var(--charcoal-color-container-hover);
}

.charcoal-radio-input:not(:disabled):active {
  background-color: var(--charcoal-color-container-press);
}

.charcoal-radio-input:not(:disabled):focus {
  outline: none;
  box-shadow: var(--charcoal-color-custom-focus-effect);
}

.charcoal-radio-input:not(:disabled):focus-visible {
  box-shadow: var(--charcoal-color-custom-focus-effect);
}

.charcoal-radio-input:not(:disabled):focus:not(:focus-visible) {
  box-shadow: none;
}

.charcoal-radio-input:not(:disabled)[aria-invalid='true'],
.charcoal-radio-input:not(:disabled)[aria-invalid='true']:focus {
  box-shadow: var(--charcoal-color-custom-error-effect);
}

.charcoal-radio-input:checked:not(:disabled):hover {
  background-color: var(--charcoal-color-container-primary-hover);
}

.charcoal-radio-input:checked:not(:disabled):hover::after {
  background-color: var(--charcoal-color-icon-on-primary-hover);
}

.charcoal-radio-input:checked:not(:disabled):active {
  background-color: var(--charcoal-color-container-primary-press);
}

.charcoal-radio-input:checked:not(:disabled):active::after {
  background-color: var(--charcoal-color-icon-on-primary-press);
}
