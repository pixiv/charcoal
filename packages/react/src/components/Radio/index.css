.charcoal-radio__label {
  display: grid;
  grid-template-columns: auto 1fr;
  grid-gap: 4px;
  align-items: center;
  cursor: pointer;
}

.charcoal-radio__label[aria-disabled]:not([aria-disabled='false']) {
  cursor: default;
  opacity: 0.32;
}

.charcoal-radio__label_div {
  font-size: 14px;
  line-height: 22px;
  color: var(--charcoal-color-text-default);
}
