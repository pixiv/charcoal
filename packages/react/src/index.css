:root {
  /* charcoal custom css variables */
  --charcoal-color-custom-focus-effect: 0 0 0 4px rgba(0, 150, 250, 0.32);
  --charcoal-color-custom-error-effect: 0 0 0 4px rgba(255, 43, 0, 0.32);
}
