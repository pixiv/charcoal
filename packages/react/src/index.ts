export {
  default as ComponentAbstraction,
  useComponentAbstraction,
  type LinkProps,
} from './core/ComponentAbstraction'
export { SSRProvider } from './core/SSRProvider'
export { default as Button, type ButtonProps } from './components/Button'
export {
  default as Clickable,
  type ClickableProps,
  type ClickableElement,
} from './components/Clickable'
export {
  default as IconButton,
  type IconButtonProps,
} from './components/IconButton'
export {
  default as Radio,
  type RadioProps,
  RadioGroup,
  type RadioGroupProps,
} from './components/Radio'
export {
  default as Select,
  type SelectProps,
  SelectGroup,
  type SelectGroupProps,
} from './components/Select'
export { default as Switch, type SwitchProps } from './components/Switch'
export {
  default as TextField,
  type TextFieldProps,
  type SingleLineTextFieldProps,
  type MultiLineTextFieldProps,
} from './components/TextField'
export { default as Icon, type IconProps } from './components/Icon'
export { default as Modal, type ModalProps } from './components/Modal'
export {
  default as LoadingSpinner,
  LoadingSpinnerIcon,
} from './components/LoadingSpinner'
export {
  default as SegmentedControl,
  type SegmentedControlProps,
} from './components/SegmentedControl'
export { default as Checkbox } from './components/Checkbox'
export { default as TagItem, type TagItemProps } from "./components/TagItem"
