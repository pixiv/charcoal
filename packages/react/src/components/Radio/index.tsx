import React, { useCallback, useContext } from 'react'
import styled from 'styled-components'
import warning from 'warning'
import { theme } from '../../styled'
import { px } from '@charcoal-ui/utils'

export type RadioProps = React.PropsWithChildren<{
  value: string
  forceChecked?: boolean
  disabled?: boolean
}>

export default function Radio({
  value,
  forceChecked = false,
  disabled = false,
  children,
}: RadioProps) {
  const {
    name,
    selected,
    disabled: isParentDisabled,
    readonly,
    hasError,
    onChange,
  } = useContext(RadioGroupContext)

  warning(
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    name !== undefined,
    `"name" is not Provided for <Radio>. Perhaps you forgot to wrap with <RadioGroup> ?`
  )

  const isSelected = value === selected
  const isDisabled = disabled || isParentDisabled
  const isReadonly = readonly && !isSelected

  const handleChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      onChange(e.currentTarget.value)
    },
    [onChange]
  )

  return (
    <RadioRoot aria-disabled={isDisabled || isReadonly}>
      <RadioInput
        name={name}
        value={value}
        checked={forceChecked || isSelected}
        hasError={hasError}
        onChange={handleChange}
        disabled={isDisabled || isReadonly}
      />
      {children != null && <RadioLabel>{children}</RadioLabel>}
    </RadioRoot>
  )
}

const RadioRoot = styled.label`
  display: grid;
  grid-template-columns: auto 1fr;
  grid-gap: ${({ theme }) => px(theme.spacing[4])};
  align-items: center;
  cursor: pointer;

  ${theme((o) => [o.disabled])}
`

export const RadioInput = styled.input.attrs({ type: 'radio' })<{
  hasError?: boolean
}>`
  /** Make prior to browser default style */
  &[type='radio'] {
    appearance: none;
    display: block;
    box-sizing: border-box;

    margin: 0;
    padding: 6px;

    width: 20px;
    height: 20px;

    ${({ hasError = false }) =>
      theme((o) => [
        o.borderRadius('oval'),
        o.bg.surface1.hover.press,
        hasError && o.outline.assertive,
      ])};

    &:not(:checked) {
      border-width: 2px;
      border-style: solid;
      border-color: ${({ theme }) => theme.color.text3};
    }

    &:checked {
      ${theme((o) => o.bg.brand.hover.press)}

      &::after {
        content: '';
        display: block;
        width: 8px;
        height: 8px;
        pointer-events: none;

        ${theme((o) => [o.bg.text5.hover.press, o.borderRadius('oval')])}
      }
    }

    ${theme((o) => o.outline.default.focus)}
  }
`

const RadioLabel = styled.div`
  ${theme((o) => [o.typography(14), o.font.text2])}
`

export type RadioGroupProps = React.PropsWithChildren<{
  className?: string
  value?: string
  label: string
  name: string
  onChange(next: string): void
  disabled?: boolean
  readonly?: boolean
  hasError?: boolean
}>

// TODO: use (or polyfill) flex gap
const StyledRadioGroup = styled.div`
  display: grid;
  grid-template-columns: 1fr;
  grid-gap: ${({ theme }) => px(theme.spacing[8])};
`

interface RadioGroupContext {
  name: string
  selected?: string
  disabled: boolean
  readonly: boolean
  hasError: boolean
  onChange: (next: string) => void
}

const RadioGroupContext = React.createContext<RadioGroupContext>({
  name: undefined as never,
  selected: undefined,
  disabled: false,
  readonly: false,
  hasError: false,
  onChange() {
    throw new Error(
      'Cannot find onChange() handler. Perhaps you forgot to wrap with <RadioGroup> ?'
    )
  },
})

export function RadioGroup({
  className,
  value,
  label,
  name,
  onChange,
  disabled,
  readonly,
  hasError,
  children,
}: RadioGroupProps) {
  const handleChange = useCallback(
    (next: string) => {
      onChange(next)
    },
    [onChange]
  )

  return (
    <RadioGroupContext.Provider
      value={{
        name,
        selected: value,
        disabled: disabled ?? false,
        readonly: readonly ?? false,
        hasError: hasError ?? false,
        onChange: handleChange,
      }}
    >
      <StyledRadioGroup
        role="radiogroup"
        aria-orientation="vertical"
        aria-label={label}
        aria-invalid={hasError}
        className={className}
      >
        {children}
      </StyledRadioGroup>
    </RadioGroupContext.Provider>
  )
}
