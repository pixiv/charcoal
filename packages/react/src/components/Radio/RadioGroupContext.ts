import * as React from 'react'

export interface RadioGroupContext {
  name: string
  selected?: string
  disabled: boolean
  readonly: boolean
  invalid: boolean
  onChange: (next: string) => void
}

export const RadioGroupContext: React.Context<RadioGroupContext> = React.createContext<RadioGroupContext>({
  name: undefined as never,
  selected: undefined,
  disabled: false,
  readonly: false,
  invalid: false,
  onChange() {
    throw new Error(
      'Cannot find onChange() handler. Perhaps you forgot to wrap with <RadioGroup> ?'
    )
  },
})
