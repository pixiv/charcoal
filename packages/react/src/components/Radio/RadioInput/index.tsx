import './index.css'

import { forwardRef, ForwardRefExoticComponent, memo, MemoExoticComponent, RefAttributes, useCallback } from 'react'
import { useClassNames } from '../../../_lib/useClassNames'

type CharcoalRadioInputProps = {
  invalid?: boolean
  onChange?: (value: string) => void
}

type InputProps = React.HTMLProps<HTMLInputElement>

export type RadioInputProps = CharcoalRadioInputProps &
  Omit<InputProps, keyof CharcoalRadioInputProps | 'ref'>

const RadioInput = forwardRef<HTMLInputElement, RadioInputProps>(
  function RadioInput({ onChange, invalid, className, ...props }, ref) {
    const handleChange = useCallback(
      (e: React.ChangeEvent<HTMLInputElement>) => {
        const el = e.currentTarget
        onChange?.(el.value)
      },
      [onChange]
    )

    const classNames = useClassNames('charcoal-radio-input', className)

    return (
      <input
        className={classNames}
        ref={ref}
        type="radio"
        onChange={handleChange}
        aria-invalid={invalid}
        {...props}
      />
    )
  }
)

const _default_1: MemoExoticComponent<ForwardRefExoticComponent<CharcoalRadioInputProps & Omit<InputProps, "ref" | keyof CharcoalRadioInputProps> & RefAttributes<HTMLInputElement>>> = memo(RadioInput)
export default _default_1
