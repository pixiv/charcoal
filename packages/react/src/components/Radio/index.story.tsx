import styled from 'styled-components'
import Radio, { RadioGroup } from '.'
import { px } from '@charcoal-ui/utils'
import { StoryObj } from '@storybook/react'
import { useState } from 'react'

export default {
  title: 'Radio',
  component: Radio,
  parameters: {
    layout: 'centered',
  },
}

const LayoutDiv = styled.div`
  display: flex;
  flex-direction: column;
  gap: ${({ theme }) => px(theme.spacing[24])};
`

export const Default: StoryObj<typeof Radio> = {
  render: function Render(args) {
    const options = ['1', '2', '3']
    const [value, setValue] = useState(options[0])

    return (
      <LayoutDiv>
        <RadioGroup
          label={'label'}
          name={'name'}
          {...args}
          value={value}
          onChange={setValue}
        >
          {options.map((option) => (
            <Radio key={option} value={option}>
              Value {option}
            </Radio>
          ))}
        </RadioGroup>
      </LayoutDiv>
    )
  },
}

export const Disabled: StoryObj<typeof Radio> = {
  render: function Render() {
    const options = ['1', '2', '3']
    const [value, setValue] = useState(options[0])
    return (
      <LayoutDiv>
        <RadioGroup
          label={'label'}
          name={'name'}
          value={value}
          onChange={setValue}
          disabled
        >
          {options.map((option) => (
            <Radio key={option} value={option} disabled>
              Value {option}
            </Radio>
          ))}
        </RadioGroup>
      </LayoutDiv>
    )
  },
}

export const PartialDisabled: StoryObj<typeof Radio> = {
  render: function Render() {
    const options = ['1', '2', '3']
    const [value, setValue] = useState(options[0])
    return (
      <LayoutDiv>
        <RadioGroup
          label={'label'}
          name={'name'}
          value={value}
          onChange={setValue}
        >
          {options.map((option) => (
            <Radio key={option} value={option} disabled={option === '2'}>
              Value {option}
            </Radio>
          ))}
        </RadioGroup>
      </LayoutDiv>
    )
  },
}

export const Readonly: StoryObj<typeof Radio> = {
  render: function Render() {
    const options = ['1', '2', '3']
    const [value, setValue] = useState(options[0])
    return (
      <LayoutDiv>
        <RadioGroup
          label={'label'}
          name={'name'}
          value={value}
          onChange={setValue}
          readonly
        >
          {options.map((option) => (
            <Radio key={option} value={option}>
              Value {option}
            </Radio>
          ))}
        </RadioGroup>
      </LayoutDiv>
    )
  },
}

export const Invalid: StoryObj<typeof Radio> = {
  render: function Render() {
    const options = ['1', '2', '3']
    const [value, setValue] = useState(options[0])
    return (
      <LayoutDiv>
        <RadioGroup
          label={'label'}
          name={'name'}
          value={value}
          onChange={setValue}
          invalid
        >
          {options.map((option) => (
            <Radio key={option} value={option}>
              Value {option}
            </Radio>
          ))}
        </RadioGroup>
      </LayoutDiv>
    )
  },
}
