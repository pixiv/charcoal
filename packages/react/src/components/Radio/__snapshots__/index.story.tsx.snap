// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Radio Default 1`] = `
.c1 {
  display: grid;
  grid-template-columns: auto 1fr;
  grid-gap: 4px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
}

.c2[type='radio'] {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  box-sizing: border-box;
  margin: 0;
  padding: 6px;
  width: 20px;
  height: 20px;
  cursor: pointer;
  border-radius: 999999px;
  background-color: var(--charcoal-surface1);
  -webkit-transition: 0.2s background-color,0.2s box-shadow;
  transition: 0.2s background-color,0.2s box-shadow;
}

.c2[type='radio']:not(:disabled):not([aria-disabled]):hover,
.c2[type='radio'][aria-disabled='false']:hover {
  background-color: var(--charcoal-surface1-hover);
}

.c2[type='radio']:not(:disabled):not([aria-disabled]):active,
.c2[type='radio'][aria-disabled='false']:active {
  background-color: var(--charcoal-surface1-press);
}

.c2[type='radio']:not(:disabled):not([aria-disabled]):focus,
.c2[type='radio'][aria-disabled='false']:focus,
.c2[type='radio']:not(:disabled):not([aria-disabled]):active,
.c2[type='radio'][aria-disabled='false']:active,
.c2[type='radio']:not(:disabled):not([aria-disabled]):focus-visible,
.c2[type='radio'][aria-disabled='false']:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c2[type='radio']:not(:checked) {
  border-width: 2px;
  border-style: solid;
  border-color: var(--charcoal-text3);
}

.c2[type='radio']:checked {
  background-color: var(--charcoal-brand);
}

.c2[type='radio']:checked::after {
  content: '';
  display: block;
  width: 8px;
  height: 8px;
  pointer-events: none;
  background-color: var(--charcoal-text5);
  border-radius: 999999px;
  -webkit-transition: 0.2s background-color,0.2s box-shadow;
  transition: 0.2s background-color,0.2s box-shadow;
}

.c2[type='radio']:checked:not(:disabled):not([aria-disabled]):hover,
.c2[type='radio']:checked[aria-disabled='false']:hover {
  background-color: var(--charcoal-brand-hover);
}

.c2[type='radio']:checked:not(:disabled):not([aria-disabled]):hover::after,
.c2[type='radio']:checked[aria-disabled='false']:hover::after {
  background-color: var(--charcoal-text5-hover);
}

.c2[type='radio']:checked:not(:disabled):not([aria-disabled]):active,
.c2[type='radio']:checked[aria-disabled='false']:active {
  background-color: var(--charcoal-brand-press);
}

.c2[type='radio']:checked:not(:disabled):not([aria-disabled]):active::after,
.c2[type='radio']:checked[aria-disabled='false']:active::after {
  background-color: var(--charcoal-text5-press);
}

.c3 {
  font-size: 14px;
  line-height: 22px;
  display: flow-root;
  color: var(--charcoal-text2);
}

.c3::before {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-top: -4px;
}

.c3::after {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-bottom: -4px;
}

.c0 {
  display: grid;
  grid-template-columns: 1fr;
  grid-gap: 8px;
}

<div
  data-dark={false}
>
  <div
    css={
      Array [
        "
      display: flex;
      flex-direction: column;
      gap: ",
        [Function],
        ";
    ",
      ]
    }
  >
    <div
      aria-invalid={false}
      aria-label="選択肢-name1"
      aria-orientation="vertical"
      className="c0"
      role="radiogroup"
    >
      <label
        aria-disabled={false}
        className="c1"
      >
        <input
          checked={false}
          className="c2"
          disabled={false}
          name="name1"
          onChange={[Function]}
          type="radio"
          value="value1"
        />
        <div
          className="c3"
        >
          name1
          (
          value1
          )を選ぶ
        </div>
      </label>
      <label
        aria-disabled={false}
        className="c1"
      >
        <input
          checked={false}
          className="c2"
          disabled={false}
          name="name1"
          onChange={[Function]}
          type="radio"
          value="value2"
        />
        <div
          className="c3"
        >
          name1
          (
          value2
          )を選ぶ
        </div>
      </label>
    </div>
    <div
      aria-invalid={false}
      aria-label="選択肢-name2"
      aria-orientation="vertical"
      className="c0"
      role="radiogroup"
    >
      <label
        aria-disabled={false}
        className="c1"
      >
        <input
          checked={false}
          className="c2"
          disabled={false}
          name="name2"
          onChange={[Function]}
          type="radio"
          value="value1"
        />
        <div
          className="c3"
        >
          name2
          (
          value1
          )を選ぶ
        </div>
      </label>
      <label
        aria-disabled={false}
        className="c1"
      >
        <input
          checked={false}
          className="c2"
          disabled={false}
          name="name2"
          onChange={[Function]}
          type="radio"
          value="value2"
        />
        <div
          className="c3"
        >
          name2
          (
          value2
          )を選ぶ
        </div>
      </label>
    </div>
    <div
      aria-invalid={false}
      aria-label="選択肢-name3"
      aria-orientation="vertical"
      className="c0"
      role="radiogroup"
    >
      <label
        aria-disabled={false}
        className="c1"
      >
        <input
          checked={false}
          className="c2"
          disabled={false}
          name="name3"
          onChange={[Function]}
          type="radio"
          value="value1"
        />
        <div
          className="c3"
        >
          name3
          (
          value1
          )を選ぶ
        </div>
      </label>
      <label
        aria-disabled={false}
        className="c1"
      >
        <input
          checked={false}
          className="c2"
          disabled={false}
          name="name3"
          onChange={[Function]}
          type="radio"
          value="value2"
        />
        <div
          className="c3"
        >
          name3
          (
          value2
          )を選ぶ
        </div>
      </label>
    </div>
  </div>
</div>
`;
